scripts/components/git_client.ts

```diff
@@ -0,0 +1,212 @@
+
+/**
+ * Client for programmatically  interacting with the local git cli
+ */
+export class GitClient {
+  private readonly gitignorePath: string;
+
+  /**
+   * execaCommand that allows us to capture stdout
+   */
+  private readonly exec;
+
+  /**
+   * execaCommand that pipes buffers to process buffers
+   */
+  private readonly execWithIO;
+
+  /**
+   * Initialize with an optional directory to operate in.
+   * Defaults to the process cwd.
+   */
+  constructor(cwd?: string) {
+    this.exec = (command: string, args: string[]) => execa.sync(command, args, { cwd });
+    this.execWithIO = (command: string, args: string[]) => execa.sync(command, args, { cwd, stdio: 'inherit' });
+    this.gitignorePath = cwd ? path.join(cwd, '.gitignore') : '.gitignore';
+  }
+
+  init = async () => {
+    await this.exec('git', ['init']);
+    await writeFile(this.gitignorePath, `node_modules${EOL}`);
+  };
+
+  /**
+   * Throws if there are uncommitted changes in the repo
+   */
+  ensureWorkingTreeIsClean = async () => {
+    const { stdout } = await this.exec('git', ['status', '--porcelain']);
+    const isDirty = stdout.trim();
+    if (isDirty) {
+      throw new Error('Dirty working tree detected. Commit or stash changes to continue.');
+    }
+  };
+
+  getCurrentBranch = async () => {
+    const { stdout: currentBranch } = await this.exec('git', ['branch', '--show-current']);
+    return currentBranch;
+  };
+
+  /**
+   * Switches to branchName. Creates the branch if it does not exist.
+   */
+  switchToBranch = async (branchName: string) => {
+    const { stdout: branchResult } = await this.exec('git', ['branch', '-l', branchName]);
+    const branchExists = branchResult.trim().length > 0;
+    if (branchExists) {
+      await this.execWithIO('git', ['switch', branchName]);
+    } else {
+      await this.execWithIO('git', ['switch', '-c', branchName]);
+    }
+  };
+
+  /**
+   * Stages and commits all current changes
+   */
+  commitAllChanges = async (message: string) => {
+    await this.execWithIO('git', ['add', '.']);
+    await this.execWithIO('git', ['commit', '--message', message, '--allow-empty']);
+  };
+
+  /**
+   * Push to the remote
+   */
+  push = async ({ force }: { force: boolean } = { force: false }) => {
+    await this.execWithIO('git', ['push', force ? '--force' : '']);
+  };
+
+  fetchTags = async () => {
+    await this.execWithIO('git', ['fetch', '--tags']);
+  };
+
+  checkout = async (ref: string, paths: string[] = []) => {
+    const additionalArgs = paths.length > 0 ? ['--', ...paths] : [];
+    await this.execWithIO('git', ['checkout', ref, ...additionalArgs]);
+  };
+
+  status = async () => {
+    await this.execWithIO('git', ['status']);
+  };
+
+  /**
+   * Returns a list of tags that point to the given commit
+   */
+  getTagsAtCommit = async (commitHash: string) => {
+    // filter out packages not published to npm
+    const privatePackages = [
```

Comment:
Not a blocker, but how can we keep this list in sync? How do we remember in the future to edit this list if we add a private package?

Reply:
I've created a `postinstall` command to generate the list of private packages so now it should always be in sync.